### DISCUSSION TITLE
이승헌 / 과제 소개

### DISCUSSION THREAD

**Discussion by heony704** [wanted-frontend-team2/pre-onboarding-8th-4-2#1] [과제 소개]
레포지토리
- https://github.com/heony704/comment-app

![image](https://user-images.githubusercontent.com/36994104/213056726-1543a5c4-85c2-4ae8-a0b3-f4b944501e4d.png)

리덕스를 처음 써봤는데 속도가 너무 느려서 아래 레포지토리 참고했습니다.
https://github.com/bezkoder/redux-toolkit-example-crud-hooks

![image](https://user-images.githubusercontent.com/36994104/213056055-5081ef9d-570b-4af5-a50b-2a210e2124dc.png)

## 패키지
- @reduxjs/toolkit
- axios
- styled-components
- @redux-devtools/extension

## 프로젝트 구조

```
src
 ┣ components
 ┃ ┣ Comment.tsx
 ┃ ┣ CommentForm.tsx
 ┃ ┣ CommentList.tsx
 ┃ ┣ InputMemo.tsx
 ┃ ┗ PageList.tsx
 ┣ pages
 ┃ ┗ CommentPage.tsx
 ┣ services
 ┃ ┣ CommentsService.ts
 ┃ ┗ http.ts
 ┣ slices
 ┃ ┣ comments.ts
 ┃ ┗ currentComment.ts
 ┣ types
 ┃ ┗ index.d.ts
 ┣ App.tsx
 ┣ index.css
 ┣ index.tsx
 ┣ reportWebVitals.ts
 ┗ store.ts
```

## api 통신

json-server와 통신하기 위한 기능은 services 폴더에 넣었습니다. 

```
services
 ┣ CommentsService.ts
 ┗ http.ts // AxiosInstance 반
```

![image](https://user-images.githubusercontent.com/36994104/213054450-6a9da50c-b38b-4748-b06b-c49462e10d84.png)

CRUD 가능합니다.

## reduxjs/toolkit 사용

createStore() 가 deprecated 라고 뜨길래 취소선 거슬려서 configureStore 써보려고 toolkit 썼습니다.
다시 찾아보니 createStore 써도 별 문제 없다는거 봐선 createStore 쓸 걸 그랬나봐요.

![image](https://user-images.githubusercontent.com/36994104/213053497-c72bcdac-e413-4bd2-b69d-4c4cc580ade3.png)

댓글 CRUD 관련 상태를 comments 에 저장했습니다.

devTools: true 도 설정해서 Redux-Devtools를 활성화했습니다.

아래는 commentReducer 입니다.

![image](https://user-images.githubusercontent.com/36994104/213054756-8e5d5fe8-e8cb-4c14-b5d9-6788aa0730c1.png)

![image](https://user-images.githubusercontent.com/36994104/213054779-99598040-45e1-4209-b4e2-2a0bb17a2f99.png)

createAsyncThunk 로 비동기 처리했습니다.

![image](https://user-images.githubusercontent.com/36994104/213054806-c7c1b139-4d3e-461f-90fe-793b991c9ba3.png)

## 못한 부분

댓글 CRUD 할 수 있는 함수는 있지만, 현재 댓글 상태를 저장할 전역 상태를 아직 만들지 못해서 
update 기능이 제대로 동작하지 않습니다.
현재 댓글 상태 저장해서 onChange 제대로 동작하면 update도 금방 될 것 같아요.

페이지네이션 기능이 없습니다.

Redux logger 진짜 모르겠어요.

**Comment by ahn0min** [comment]
오.. extraReducers 좋아보이네요..
redux-saga vs redux-thunk = thunk..?
